{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d04fece6-28c8-4497-bfbb-aa74b9930bd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copyright (c) OpenMMLab. All rights reserved.\n",
    "import pickle\n",
    "import argparse\n",
    "\n",
    "import cv2\n",
    "import mmcv\n",
    "import mmengine\n",
    "import torch\n",
    "from mmengine import DictAction\n",
    "from mmengine.utils import track_iter_progress\n",
    "\n",
    "from mmaction.apis import (detection_inference,\n",
    "                           init_recognizer, pose_inference)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c986ffe-8454-493d-beec-3393b035b3dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def frame_extraction(video_path, short_side):\n",
    "    \"\"\"Extract frames given video_path.\n",
    "\n",
    "    Args:\n",
    "        video_path (str): The video_path.\n",
    "    \"\"\"\n",
    "    # Load the video, extract frames into ./tmp/video_name\n",
    "    target_dir = osp.join('./tmp', osp.splitext(video_path)[0])\n",
    "    os.makedirs(target_dir, exist_ok=True)\n",
    "    # Should be able to handle videos up to several hours\n",
    "    frame_tmpl = osp.join(target_dir, 'img_{:06d}.jpg')\n",
    "    vid = cv2.VideoCapture(video_path)\n",
    "    frames = []\n",
    "    frame_paths = []\n",
    "    flag, frame = vid.read()\n",
    "    cnt = 0\n",
    "    new_h, new_w = None, None\n",
    "    while flag:\n",
    "        if new_h is None:\n",
    "            h, w, _ = frame.shape\n",
    "            new_w, new_h = mmcv.rescale_size((w, h), (short_side, np.Inf))\n",
    "\n",
    "        frame = mmcv.imresize(frame, (new_w, new_h))\n",
    "\n",
    "        frames.append(frame)\n",
    "        frame_path = frame_tmpl.format(cnt + 1)\n",
    "        frame_paths.append(frame_path)\n",
    "\n",
    "        cv2.imwrite(frame_path, frame)\n",
    "        cnt += 1\n",
    "        flag, frame = vid.read()\n",
    "\n",
    "    return frame_paths, frames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4528bc61-b248-483b-9e58-c2f220d9f25a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_args():\n",
    "    parser = argparse.ArgumentParser(description='MMAction2 demo')\n",
    "    # parser.add_argument('video', help='video file/url')\n",
    "    # parser.add_argument('out_filename', help='output filename')\n",
    "\n",
    "    parser.add_argument(\n",
    "        '--det-config',\n",
    "        default='demo/demo_configs/faster-rcnn_r50_fpn_2x_coco_infer.py',\n",
    "        help='human detection config file path (from mmdet)')\n",
    "    parser.add_argument(\n",
    "        '--det-checkpoint',\n",
    "        default=('http://download.openmmlab.com/mmdetection/v2.0/faster_rcnn/'\n",
    "                 'faster_rcnn_r50_fpn_2x_coco/'\n",
    "                 'faster_rcnn_r50_fpn_2x_coco_'\n",
    "                 'bbox_mAP-0.384_20200504_210434-a5d8aa15.pth'),\n",
    "        help='human detection checkpoint file/url')\n",
    "    parser.add_argument(\n",
    "        '--det-score-thr',\n",
    "        type=float,\n",
    "        default=0.9,\n",
    "        help='the threshold of human detection score')\n",
    "    parser.add_argument(\n",
    "        '--det-cat-id',\n",
    "        type=int,\n",
    "        default=0,\n",
    "        help='the category id for human detection')\n",
    "    \n",
    "    parser.add_argument(\n",
    "        '--pose-config',\n",
    "        default='demo/demo_configs/'\n",
    "        'td-hm_hrnet-w32_8xb64-210e_coco-256x192_infer.py',\n",
    "        help='human pose estimation config file path (from mmpose)')\n",
    "    parser.add_argument(\n",
    "        '--pose-checkpoint',\n",
    "        default=('https://download.openmmlab.com/mmpose/top_down/hrnet/'\n",
    "                 'hrnet_w32_coco_256x192-c78dce93_20200708.pth'),\n",
    "        help='human pose estimation checkpoint file/url')\n",
    "    parser.add_argument(\n",
    "        '--label-map',\n",
    "        default='tools/data/skeleton/label_map_ntu60.txt',\n",
    "        help='label map file')\n",
    "    parser.add_argument(\n",
    "        '--device', type=str, default='cuda:0', help='CPU/CUDA device option')\n",
    "    \n",
    "\n",
    "    args = parser.parse_args()\n",
    "    return args"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f4e91855-8157-4e79-962f-945273244306",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(video_idx):\n",
    "    video_path = file_paths[video_idx]\n",
    "    args = parse_args()\n",
    "    # video_path = args.video\n",
    "    frame_paths, original_frames = frame_extraction(video_path,\n",
    "                                                    args.short_side)\n",
    "    num_frame = len(frame_paths)\n",
    "    h, w, _ = original_frames[0].shape\n",
    "\n",
    "    tmp_frame_dir = osp.dirname(frame_paths[0])\n",
    "    print(f\"tmp_frame_dir: {tmp_frame_dir}\")\n",
    "\n",
    "    # Get Human detection results\n",
    "    det_results = detection_inference(args, frame_paths)\n",
    "    torch.cuda.empty_cache()\n",
    "\n",
    "    pose_results = pose_inference(args, frame_paths, det_results)\n",
    "    torch.cuda.empty_cache()\n",
    "\n",
    "    fake_anno = dict(\n",
    "        frame_dir='',\n",
    "        label=-1,\n",
    "        img_shape=(h, w),\n",
    "        original_shape=(h, w),\n",
    "        start_index=0,\n",
    "        modality='Pose',\n",
    "        total_frames=num_frame)\n",
    "\n",
    "    num_person = max([len(x) for x in pose_results])\n",
    "    # Current PoseC3D models are trained on COCO-keypoints (17 keypoints)\n",
    "    num_keypoint = 17\n",
    "    keypoint = np.zeros((num_person, num_frame, num_keypoint, 2),\n",
    "                        dtype=np.float16)\n",
    "    keypoint_score = np.zeros((num_person, num_frame, num_keypoint),\n",
    "                              dtype=np.float16)\n",
    "    for i, poses in enumerate(pose_results):\n",
    "        for j, pose in enumerate(poses):\n",
    "            pose = pose['keypoints']\n",
    "            keypoint[j, i] = pose[:, :2]\n",
    "            keypoint_score[j, i] = pose[:, 2]\n",
    "    fake_anno['keypoint'] = keypoint\n",
    "    fake_anno['keypoint_score'] = keypoint_score\n",
    "\n",
    "       # save fake_anno to pkl\n",
    "    with open(f'{tmp_frame_dir}/anno.pkl', 'wb') as f:\n",
    "        pickle.dump(fake_anno, f)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0557da24-8c9c-412d-9b0a-93faca44f467",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os\n",
    "\n",
    "# List of folder names from 1 to 17\n",
    "folder_names = [str(i) for i in range(0, 16)]\n",
    "\n",
    "# Initialize an empty list to store the file paths\n",
    "file_paths = []\n",
    "\n",
    "# Iterate over each folder\n",
    "for folder_name in folder_names:\n",
    "    folder_path = f\"SetA1_clip/{folder_name}\"\n",
    "    if os.path.isdir(folder_path):\n",
    "        # Walk through the folder directory tree\n",
    "        for root, dirs, files in os.walk(folder_path):\n",
    "            for file in files:\n",
    "                file_path = os.path.join(root, file)\n",
    "                file_paths.append(file_path)\n",
    "\n",
    "# # Print the file paths\n",
    "# for file_path in file_paths:\n",
    "#     print(file_path)\n",
    "#     main(file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "03f95c0e-526c-475d-a493-9859fec86aab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4644"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(file_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "39208ac3-0040-450d-b468-cd7c530395e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a8d302ef-71b7-46dd-83ff-50b1b7a91828",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/datnt114/anaconda3/envs/openmmlab/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "from tqdm import tqdm, auto\n",
    "from multiprocessing import pool\n",
    "num_files = len(file_paths)\n",
    "start_idx = 0\n",
    "end_idx = num_files\n",
    "results = pool.ThreadPool(2).imap(main, list(range(start_idx, end_idx)))\n",
    "pbar = auto.tqdm(results, total=end_idx - start_idx)\n",
    "for _ in pbar:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8f17804e-25fb-46ca-ae89-bd913ef83225",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-f93025cf-6288-4d70-8f1c-8dc583c06587.json\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001B[0;31mSystemExit\u001B[0m\u001B[0;31m:\u001B[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/datnt114/.local/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3516: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "main(0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "67df0ad6-3ef4-4c80-9261-9e910dded8e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "  0%|                                                                | 0/4644 [00:00<?, ?it/s]ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "usage: ipykernel_launcher.py [-h] [--det-config DET_CONFIG]\n",
      "                             [--det-checkpoint DET_CHECKPOINT]\n",
      "                             [--det-score-thr DET_SCORE_THR]\n",
      "                             [--det-cat-id DET_CAT_ID]\n",
      "                             [--pose-config POSE_CONFIG]\n",
      "                             [--pose-checkpoint POSE_CHECKPOINT]\n",
      "                             [--label-map LABEL_MAP] [--device DEVICE]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /home/datnt114/.local/share/jupyter/runtime/kernel-247ad608-df91-4371-8748-603cff8448db.json\n",
      "  0%|                                                                | 0/4644 [01:11<?, ?it/s]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mIndexError\u001B[0m                                Traceback (most recent call last)",
      "File \u001B[0;32m~/anaconda3/envs/openmmlab/lib/python3.10/multiprocessing/pool.py:856\u001B[0m, in \u001B[0;36mIMapIterator.next\u001B[0;34m(self, timeout)\u001B[0m\n\u001B[1;32m    855\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m--> 856\u001B[0m     item \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_items\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mpopleft\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    857\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mIndexError\u001B[39;00m:\n",
      "\u001B[0;31mIndexError\u001B[0m: pop from an empty deque",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001B[0;31mKeyboardInterrupt\u001B[0m                         Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[9], line 3\u001B[0m\n\u001B[1;32m      1\u001B[0m results \u001B[38;5;241m=\u001B[39m pool\u001B[38;5;241m.\u001B[39mThreadPool(\u001B[38;5;241m16\u001B[39m)\u001B[38;5;241m.\u001B[39mimap(main, \u001B[38;5;28mlist\u001B[39m(\u001B[38;5;28mrange\u001B[39m(start_idx, end_idx)))\n\u001B[1;32m      2\u001B[0m pbar \u001B[38;5;241m=\u001B[39m auto\u001B[38;5;241m.\u001B[39mtqdm(results, total\u001B[38;5;241m=\u001B[39mend_idx \u001B[38;5;241m-\u001B[39m start_idx)\n\u001B[0;32m----> 3\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m _ \u001B[38;5;129;01min\u001B[39;00m pbar:\n\u001B[1;32m      4\u001B[0m     \u001B[38;5;28;01mpass\u001B[39;00m\n",
      "File \u001B[0;32m~/anaconda3/envs/openmmlab/lib/python3.10/site-packages/tqdm/std.py:1182\u001B[0m, in \u001B[0;36mtqdm.__iter__\u001B[0;34m(self)\u001B[0m\n\u001B[1;32m   1179\u001B[0m time \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_time\n\u001B[1;32m   1181\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m-> 1182\u001B[0m     \u001B[38;5;28;01mfor\u001B[39;00m obj \u001B[38;5;129;01min\u001B[39;00m iterable:\n\u001B[1;32m   1183\u001B[0m         \u001B[38;5;28;01myield\u001B[39;00m obj\n\u001B[1;32m   1184\u001B[0m         \u001B[38;5;66;03m# Update and possibly print the progressbar.\u001B[39;00m\n\u001B[1;32m   1185\u001B[0m         \u001B[38;5;66;03m# Note: does not call self.update(1) for speed optimisation.\u001B[39;00m\n",
      "File \u001B[0;32m~/anaconda3/envs/openmmlab/lib/python3.10/multiprocessing/pool.py:861\u001B[0m, in \u001B[0;36mIMapIterator.next\u001B[0;34m(self, timeout)\u001B[0m\n\u001B[1;32m    859\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_pool \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[1;32m    860\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mStopIteration\u001B[39;00m \u001B[38;5;28;01mfrom\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m\n\u001B[0;32m--> 861\u001B[0m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_cond\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mwait\u001B[49m\u001B[43m(\u001B[49m\u001B[43mtimeout\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    862\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[1;32m    863\u001B[0m     item \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_items\u001B[38;5;241m.\u001B[39mpopleft()\n",
      "File \u001B[0;32m~/anaconda3/envs/openmmlab/lib/python3.10/threading.py:320\u001B[0m, in \u001B[0;36mCondition.wait\u001B[0;34m(self, timeout)\u001B[0m\n\u001B[1;32m    318\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:    \u001B[38;5;66;03m# restore state no matter what (e.g., KeyboardInterrupt)\u001B[39;00m\n\u001B[1;32m    319\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m timeout \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[0;32m--> 320\u001B[0m         \u001B[43mwaiter\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43macquire\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    321\u001B[0m         gotit \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mTrue\u001B[39;00m\n\u001B[1;32m    322\u001B[0m     \u001B[38;5;28;01melse\u001B[39;00m:\n",
      "\u001B[0;31mKeyboardInterrupt\u001B[0m: "
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "943e6f3a-549b-4f80-b293-d0fbafd6c554",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
